### PR TITLE
Add documentation for using Ray Tune for Horovod

### PR DESCRIPTION
**Pull request by richardliaw**
## Checklist before submitting

- [ ] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs?
- [ ] Did you write any tests to validate this change?  
- [ ] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes # (issue).

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

File: docs/hyperparameter_search.rst [hunk 1]
```diff
@@ -0,0 +1,136 @@
+.. inclusion-marker-start-do-not-remove
+Distributed Hyperparameter Search
+=================================
+
+Horovod's data parallelism training capabilities allows you to scale out and speed up the workload of training a deep learning model. However, distributed training often does not exhibit linear scaling in accuracy improvement - i.e., simply using 2x more workers does not necessarily mean the model will obtain the same accuracy in 2x less time.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:5]: allows --> allow

Comment (sblotner) [inline docs/hyperparameter_search.rst:5]: "improvement - i.e.,"  --> "improvement, e.g., "

File: docs/hyperparameter_search.rst [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+To address this, you will often need to re-tune hyperparameters when training at scale, as many hyperparameters exhibit different behaviors at larger scales.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:7]: you will often --> you often

File: docs/hyperparameter_search.rst [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+To address this issue, Horovod offers a Ray Tune integration to enable parallel hyperparameter tuning with distributed training.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:9]: Can remove "To address this issue,"

File: docs/hyperparameter_search.rst [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+Ray Tune is an industry standard tool for distributed hyperparameter tuning. Ray Tune includes the latest hyperparameter search algorithms, integrates with TensorBoard and other analysis libraries, and natively supports distributed training. The Ray Tune + Horovod integration leverages the underlying Ray framework to provide a scalable and comprehensive hyperparameter tuning setup.
+
+**By the end of this guide, you will learn:**
+* How you can set up Ray Tune and Horovod to tune your hyperparameters
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:14]: I think you need a blank line before the list begins

File: docs/hyperparameter_search.rst [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            time.sleep(1)
+            tune.report(test=1, rank=hvd.rank())
+
+    trainable = DistributedTrainableCreator(
+            training_function, num_slots=2, use_gpu=use_gpu)
+    analysis = tune.run(
+            trainable, num_samples=2, config={"epochs": tune.grid_search([1, 2, 3]))
+    print(analysis.best_config)
+
+Not only is the interface very simple, it is also packed with features.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:40]: I recommend removing this line. We generally recommend avoiding calling something simple in technical writing to reduce assumptions about reader knowledge/frustration

File: docs/hyperparameter_search.rst [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+Basic setup
+-----------
+
+Use the `DistributedTrainableCreator`_ function to adapt your Horovod training function to be compatible with Tune.
+
+`DistributedTrainableCreator`_ exposes ``num_hosts``, ``num_slots``, ``use_gpu`` and ``num_cpus_per_slot`` - these parameters allow you to specify the resource allocation of a single "trial" (or "Trainable") which itself can be a distributed training job.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:47]: add comma after "use_gpu"

Comment (sblotner) [inline docs/hyperparameter_search.rst:47]: "- these parameters allow you to specify" --> ". Use these parameters to specify"

File: docs/hyperparameter_search.rst [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+    trainable = DistributedTrainableCreator(
+        training_function, num_slots=2, use_gpu=True)
+
+The training function itself must do three things:
+
+1. It must adhere to the Tune Function API signature.
+1. Its body must include a ``horovod.init()`` call.
+1. It must call ``tune.report`` during training, typically called iteratively at the end of every epoch.
+
+Setting up a tuning cluster
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:62]: Setting --> Set

File: docs/hyperparameter_search.rst [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+---------------------------
+
+You can easily leverage Ray Tune with Horovod on a laptop, single machine with multiple GPUs, or across multiple machines. To run on a single machine, you can execute your python script as-is (for example, horovod_simple.py, assuming Ray and Horovod are installed properly):
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:65]: remove "easily"

Comment (sblotner) [inline docs/hyperparameter_search.rst:65]: remove "you can"
python --> Python

File: docs/hyperparameter_search.rst [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+.. code-block:: bash
+
+    python horovod_simple.py
+
+
+To leverage a distributed hyperparameter tuning setup with Ray Tune + Horovod, you’ll need to install Ray and set up a Ray cluster. Ray clusters are started with the Ray Cluster Launcher or manually.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:72]: remove "you'll need to"

Comment (sblotner) [inline docs/hyperparameter_search.rst:72]: remove "you'll need to"

Comment (sblotner) [inline docs/hyperparameter_search.rst:72]: Ray clusters are started --> Start a Ray cluster

File: docs/hyperparameter_search.rst [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+Below, we’ll use the Ray cluster launcher, but you can start Ray on any list of nodes, on any cluster manager or cloud provider.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:74]: Cluster Launcher

File: docs/hyperparameter_search.rst [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+To use the cluster launcher, you’ll first want to specify a configuration file. Below we have an example of using AWS EC2, but you can easily launch the cluster on any cloud provider:
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:76]: Cluster Launcher

"you'll first want to specify" --> "first specify"

Comment (sblotner) [inline docs/hyperparameter_search.rst:76]: remove "easily"

File: docs/hyperparameter_search.rst [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+    min_workers: 3
+    max_workers: 3
+
+    # Deep Learning AMI (Ubuntu) Version 21.0
+    head_node: {InstanceType: p3.2xlarge, ImageId: ami-0b294f219d14e6a82}
+    worker_nodes: {InstanceType: p3.2xlarge, ImageId: ami-0b294f219d14e6a82}
+    setup_commands: # Set up each node.
+        - HOROVOD_WITH_GLOO=1 HOROVOD_GPU_OPERATIONS=NCCL pip install horovod[ray]
+
+You can then run ray up ray_cluster.yaml, and a cluster of 4 nodes will be automatically started for you with Ray.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:93]: Is ray up ray_cluster.yaml the command? If so, recommend making this:
`ray up ray_cluster.yaml`

**Review by sblotner — COMMENTED** [review]
Nice docs! I've added some comments with suggested editorial changes

**Comment by richardliaw** [comment]
Hi @sblotner ! Thanks a bunch for the thorough review. I haven't actually pushed the entire draft yet; will it be possible to review after the PR is marked as ready?

**Comment by sblotner** [comment]
> Hi @sblotner ! Thanks a bunch for the thorough review. I haven't actually pushed the entire draft yet; will it be possible to review after the PR is marked as ready?

Definitely! Will review the PR end-to-end once it's ready

File: docs/hyperparameter_search.rst (hunk 5 above)
Comment (richardliaw) [reply docs/hyperparameter_search.rst:53]: that's a great point, thanks for the tip!

File: docs/hyperparameter_search.rst (hunk 12 above)
Comment (richardliaw) [reply docs/hyperparameter_search.rst]: done!

File: docs/hyperparameter_search.rst (hunk 10 above)
Comment (richardliaw) [reply docs/hyperparameter_search.rst]: done!

File: docs/hyperparameter_search.rst (hunk 9 above)
Comment (richardliaw) [reply docs/hyperparameter_search.rst]: done!

File: docs/hyperparameter_search.rst (hunk 4 above)
Comment (richardliaw) [reply docs/hyperparameter_search.rst]: done!

File: docs/hyperparameter_search.rst (hunk 6 above)
Comment (richardliaw) [reply docs/hyperparameter_search.rst:60]: done!

File: docs/hyperparameter_search.rst (hunk 9 above)
Comment (richardliaw) [reply docs/hyperparameter_search.rst]: done!

Comment (richardliaw) [reply docs/hyperparameter_search.rst]: done!

**Comment by richardliaw** [comment]
@sblotner thanks a bunch for these suggestions - they're really really helpful! I've made adjustments and pushed the rest of the document. 

Could you take a look?

File: docs/hyperparameter_search.rst [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+2. Its body must include a ``horovod.init()`` call.
+3. It must call ``tune.report`` (`docs <https://docs.ray.io/en/latest/tune/api_docs/trainable.html#tune-report-tune-checkpoint-function-api>`__) during training, typically called iteratively at the end of every epoch.
+
+
+Optimization of hyperparameters
+-------------------------------
+
+`Ray Tune`_ is able to orchestrate complex computational patterns with the `Ray Actor API <https://docs.ray.io/en/latest/actors.html>`__. For hyperparameter tuning, `Ray Tune`_ is able to conduct `parallel bayesian optimization <https://docs.ray.io/en/latest/tune/api_docs/suggestion.html>`__ and `Population Based Training <https://docs.ray.io/en/latest/tune/api_docs/schedulers.html>`__ on a group of distributed models.
+
+The user may need to implement model checkpointing. The rest of the optimization process can be configured with just a couple lines of code.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:79]: The user --> You

remove "just"

File: docs/hyperparameter_search.rst [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+      Monitor autoscaling with
+        ray exec ~/dev/cfgs/check-autoscaler.yaml 'tail -n 100 -f /tmp/ray/session_latest/logs/monitor*'
+      Connect to a terminal on the cluster head:
+        ray attach ~/dev/cfgs/check-autoscaler.yaml
+      Get a remote shell to the cluster manually:
+        ssh -o IdentitiesOnly=yes -i ~/.ssh/ray-autoscaler_2_us-west-2.pem ubuntu@34.217.192.11
+
+After the cluster is up, you can ssh into the head node and run your Tune script there.
+
+Implementation (Underneath the hood)
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:218]: Underneath --> underneath

File: docs/hyperparameter_search.rst [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+------------------------------------
+
+Underneath the hood, `Ray Tune`_ will launch multiple "`trials <https://docs.ray.io/en/latest/tune/key-concepts.html#tune-run-and-trials>`__" in parallel. Each of these trials reference a `set of Ray actors <https://docs.ray.io/en/latest/actors.html>`__. For each trial, there will be 1 “coordinator actor”, and this coordinator actor will manage N training actors. One basic assumption of this implementation is that all sub-workers of a trial will be placed evenly across different machines.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:221]: "coordinator actor", --> "coordinator actor,"

File: docs/hyperparameter_search.rst [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+
+.. image:: media/tune-horovod.jpg
+
+Training actors will each hold a copy of the model and will create a communication group for Horovod allreduce. Training will execute on each actor, reporting intermediate metrics back to Tune.
+
+This API requires gloo as the underlying communication primitive. Be sure to install Horovod with ``HOROVOD_WITH_GLOO`` `enabled <https://horovod.readthedocs.io/en/stable/install_include.html#gloo>`__.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:227]: gloo --> Gloo

File: docs/hyperparameter_search.rst [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+* Linear scaling of learning rates: When the minibatch size is multiplied by k, multiply the learning rate by k.
+* Dynamically adjusting batch size over the course of training:
+
+  - One of the original papers presents a simple baseline of increasing the batch size over time
+  - `ABSA provides a way <https://openreview.net/pdf?id=H1lnJ2Rqt7>`__ to leverage second order information to guide the batch size over time
+  - `Gradient noise scale <https://openai.com/blog/science-of-ai/>`__ can be calculated to guide the increase of batch size over time
+
+To leverage a dynamically changing batch size in training, you should either:
+
+* leverage `gradient accumulation <https://gist.github.com/thomwolf/ac7a7da6b1888c2eeac8ac8b9b05d3d3>`__.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:255]: leverage --> Leverage

Comment (sblotner) [inline docs/hyperparameter_search.rst:255]: can remove the period at the end of this line

File: docs/hyperparameter_search.rst [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+* Implement your own TrialScheduler to dynamically change the number of workers (coming soon)
+
+Parameter: Learning rate schedules (warmup)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As noted in this `Facebook Imagenet Training paper <https://research.fb.com/wp-content/uploads/2017/06/imagenet1kin1h5.pdf>`__, the linear scaling rule breaks down when the network is rapidly changing, which commonly occurs in early stages of training. This issue can be addressed with a “warmup”, which is a strategy of using less aggressive learning rates at the start of training.
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:261]: "warmup", --> "warmup,"

File: docs/hyperparameter_search.rst [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+
+**What are common solutions?**
+
+Goyal et al. (2017) proposes a warm-up schedule, where training usually starts with a small learning rate and gradually increased to match a larger target learning rate. After the warm-up period (usually a few epochs), a regular learning rate schedule is used ("multi-steps", polynomial decay etc). Thus, there are generally three parameters for warmup schedule:
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:265]: rate and gradually --> rate, gradually

File: docs/hyperparameter_search.rst [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+
+* length of warmup (number of epochs)
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:267]: length --> Length

File: docs/hyperparameter_search.rst [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+* starting learning rate
```
Comment (sblotner) [inline docs/hyperparameter_search.rst:268]: starting --> Starting

**Review by sblotner — COMMENTED** [review]
This looks great! Added some more editorial suggestions to consider

**Comment by richardliaw** [comment]
@sblotner did another pass!

**Review by sblotner — APPROVED** [review]
Looks great! Thanks for making all of those revisions. 

My only remaining suggestion would be to reduce the number of times you link to https://docs.ray.io/en/latest/tune/ when you say "Ray Tune" (you can probably just mention it once in the beginning of the doc or only where you expect the reader to click into the link).

**Review by tgaddair — APPROVED** [review]
LGTM!  Nicely done.

**Pull request merged**